README.md
# Ceph Rados Block Device Docker VolumeDriver Plugin

* Use Case: Persistent Storage for a Single Docker Container
  * an RBD Image can only be used by 1 Docker Container at a time

* Plugin is a separate process running alongside Docker Daemon
  * plugin can only be configured for a single Ceph User (currently no 
    good way to pass things from docker -> plugin, get only volume name)
  * run multiple plugin instances for varying configs (ceph user, default pool, default size)
  * can pass extra config via volume name to override default pool and creation size:
    * docker run --volume-driver rbd -v poolname/imagename@size:/mnt/disk1 ...

* plugin supports all Docker VolumeDriver Plugin API commands:
  * Create - can provision Ceph RBD Image in a pool of a certain size
    * controlled by `--create` boolean flag (default false)
    * default size from `--size` flag (default 20480 = 20GB)
  * Mount - Locks, Maps and Mounts RBD Image to the Host system
  * Unmount - Unmounts, Unmaps and Unlocks the RBD Image on request
  * Remove - Removes (destroys) RBD Image on request
    * only called for `docker run --rm -v ...` or `docker rm -v ...`
    * action controlled by plugin's `--remove` flag, which can be one of three values:
      - ''ignore'' - the call to delete the ceph rbd volume is ignored (default)
      - ''rename'' - will cause image to be renamed with _zz_ prefix for later culling
      - ''delete'' - will actually delete ceph rbd image (destructive)
  * Get, List

* for there is problems in "exclusive-lock" feature, it must be disabled


* this plugin must run before docker daemon

## user of this plugin must apply patch ceph-patch/rbd-nbd-ceph-10.2.2.patch

we add lock in do\_map function.

## Generate RPMs

	$ sudo yum install -y rpm-build
	$ make rpms

## Plugin Setup

Plugin is a standalone process and places a Socket file in a known 
location.  Generally need to start this before running Docker.  It does 
not daemonize as it is expected to be controlled by systemd, so if you 
need it in the background, use normal shell process control (&).

The socket has a name (sans .sock) which is used to refer to the plugin
via the `--volume-driver=name` docker CLI option, allowing multiple
uniquely named plugin instances with different default configurations.

The default name for the socket is "rbd", so you would refer to 
`--volume-driver rbd` from docker.

General build/run requirements:
* librados2-devel and librbd1-devel for go-ceph
* /usr/bin/rbd for mapping and unmapping to kernel
* /usr/sbin/mkfs.xfs for fs creation
* /usr/bin/mount and /usr/bin/umount

Tested with Ceph version 10.2.2 on Centos 7.1 host with Docker 1.12.6 
and docker api version 1.24.

### Commandline Options

	Usage of rbd-docker-plugin:
	  -cluster string
	        xtao ceph cluster (default "xtao")
	  -config string
	        Xtao ceph cluster config (default "/etc/ceph/xtao.conf")
	  -create
	        Can auto Create RBD Images (default true)
	  -debug
	        Debug output
	  -fs string
	        FS type for the created RBD Image (must be xfs now) (default "xfs")
	  -go-ceph
	        Use go-ceph library
	  -logdir string
	        Logfile directory (default "/var/log")
	  -mount string
	        Mount directory for volumes on host (default "/var/lib/docker-volumes")
	  -name string
	        Docker plugin name for use on --volume-driver option (default "rbd")
	  -plugins string
	        Docker plugin directory for socket (default "/run/docker/plugins")
	  -pool string
	        Default Ceph Pool for RBD operations (default "rbd")
	  -remove value
	        Action to take on Remove: ignore, delete or rename (default ignore)
	  -size int
	        RBD Image size to Create (in MB) (default: 20480=20GB) (default 20480)
	  -use-nbd
	        Use rbd-nbd to map RBD Image (default true)
	  -user string
	        Ceph user (default "admin")
	  -version
	        Print version
	
### Running Plugin

Start with the default options:

* socket name=rbd, pool=rbd, user=admin, logfile=/var/log/rbd-docker-plugin.log
* no creation or removal of volumes

    sudo rbd-docker-plugin
    # docker run --volume-driver rbd -v ...

For Debugging: send log to STDERR:

    sudo RBD_DOCKER_PLUGIN_DEBUG=1 rbd-docker-plugin

Use a different socket name and Ceph pool

    sudo rbd-docker-plugin --name rbd2 --pool liverpool
    # docker run --volume-driver rbd2 -v ...

To allow creation of new RBD Images:

    sudo rbd-docker-plugin --create

To allow creation and removal:

    sudo rbd-docker-plugin --create --remove

### Testing

Use with docker 1.8+ which has the `--volume-driver` support.

* https://docker.com/

Alternatively, you can POST json to the socket to test the functionality
manually.  If your curl is new enough (v7.40+), you can use the
`--unix-socket` option and syntax.  You can also use [this golang
version](https://github.com/Soulou/curl-unix-socket) instead:

    go get github.com/Soulou/curl-unix-socket


Once you have that you can POST json to the plugin:

    % sudo curl-unix-socket -v -X POST unix:///run/docker/plugins/rbd.sock:/Plugin.Activate
    > POST /Plugin.Activate HTTP/1.1
    > Socket: /run/docker/plugins/rbd.sock
    > Content-Length: 0
    >
    < HTTP/1.1 200 OK
    < Content-Type: appplication/vnd.docker.plugins.v1+json
    < Date: Tue, 28 Jul 2015 18:52:11 GMT
    < Content-Length: 33
    {"Implements": ["VolumeDriver"]}


    # Plugin started without --create:
    % sudo curl-unix-socket -v -X POST -d '{"Name": "testimage"}' unix:///run/docker/plugins/rbd.sock:/VolumeDriver.Create
    > POST /VolumeDriver.Create HTTP/1.1
    > Socket: /run/docker/plugins/rbd.sock
    > Content-Length: 21
    >
    < HTTP/1.1 500 Internal Server Error
    < Content-Length: 62
    < Content-Type: appplication/vnd.docker.plugins.v1+json
    < Date: Tue, 28 Jul 2015 18:53:20 GMT
    {"Mountpoint":"","Err":"Ceph RBD Image not found: testimage"}

    # Plugin started --create turned on will create unknown image:
    % sudo curl-unix-socket -v -X POST -d '{"Name": "testimage"}' unix:///run/docker/plugins/rbd.sock:/VolumeDriver.Create
    > POST /VolumeDriver.Create HTTP/1.1
    > Socket: /run/docker/plugins/rbd.sock
    > Content-Length: 21
    >
    < HTTP/1.1 200 OK
    < Content-Length: 27
    < Content-Type: appplication/vnd.docker.plugins.v1+json
    < Date: Fri, 14 Aug 2015 19:47:35 GMT
    {"Mountpoint":"","Err":""}

## Examples

If you need persistent storage for your application container, you can use a Ceph Block Device as a persistent disk.

You can provision the Block Device and Filesystem first, or allow a
sufficiently configured Plugin instance create it for you.
This plugin can create RBD images with XFS filesystem.

1. (Optional) Provision RBD Storage yourself
  * `sudo rbd create --size 1024 foo`
  * `sudo rbd map foo`  => /dev/rbd1
  * `sudo mkfs.xfs /dev/rbd1`
  * `sudo rbd unmap /dev/rbd1`
2. Or Run the RBD Docker Plugin with `--create` option flag
  * `sudo rbd-docker-plugin --create`
3. Requesting and Using Storage
  * `docker run --volume-driver=rbd --volume foo:/mnt/foo -it ubuntu /bin/bash`
  * Volume will be locked, mapped and mounted to Host and bind-mounted to container at `/mnt/foo`
  * When container exits, the volume will be unmounted, unmapped and unlocked
  * You can control the RBD Pool and initial Size using this syntax:
    * foo@1024 => pool=rbd (default), image=foo, size 1GB
    * deep/foo =>  pool=deep, image=foo and default `--size` (20GB)
    * deep/foo@1024 => pool=deep, image=foo, size 1GB
    - pool must already exist

### Misc

* RBD Snapshots: `sudo rbd snap create --image foo --snap foosnap`
* Resize RBD image:
  * set max size: `sudo rbd resize --size 2048 --image foo`
  * map/mount and then fix XFS: `sudo xfs_growfs -d /mnt/foo`

## TODO

* add cluster config options to support non-default clusters
* figure out how to test
  * do we need a working ceph cluster?
  * docker containers with ceph?

## Links

### Docker API Documentation

- [Experimental: Extend Docker with a Plugin](https://github.com/docker/docker/blob/master/experimental/plugins.md)
- [Experimental: Docker Plugin API](https://github.com/docker/docker/blob/master/experimental/plugin_api.md)
- [Experimental: Docker Volume Plugins](https://github.com/docker/docker/blob/master/experimental/plugins_volume.md)

### Code Examples and Libraries

- https://plugins-demo-2015.github.io
- [Flocker Docker Plugin](https://docs.clusterhq.com/en/1.0.3/labs/docker-plugin.html)
- VolumeDriver golang framework: https://github.com/docker/go-plugins-helpers/tree/master/volume
- GlusterFS Example: https://github.com/calavera/docker-volume-glusterfs
- KeyWhiz example: https://github.com/calavera/docker-volume-keywhiz
- Ceph Rados, RBD golang lib: https://github.com/ceph/go-ceph

Related Projects
- https://github.com/AcalephStorage/docker-volume-ceph-rbd
- https://github.com/contiv/volplugin

# Packaging

Using [tpkg](http://tpkg.github.io) package to distribute and specify 
native package dependencies.  Tested with Centos 7.1 and yum/rpm 
packages.


# License

This project is using the MIT License (MIT), see LICENSE file.

Copyright (c) 2015 YP LLC


CHANGELOG.md
# Change Log
All notable changes to project should be documented in this file.
We attempt to adhere to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added
### Removed
### Changed

## [2.0.2] - 2017-08-28
### Changed
- Revert to previous Unmount behavior that would attempt a limited
  unmount in the case volume is not in memory array. This can also occur
when plugin is restarted so we want to unmount volumes started with old
plugin

## [2.0.1] - 2017-08-28
### Changed
- fixed List and Get methods to only return Mountpoint value if volume is
actually mounted.  Hopefully this matches the Docker Volume API better and
fixes some spurious volume creation (See #3, #6, #12, and
yp-engineering/rbd-docker-plugin#36)

## [2.0.0] - 2017-08-25
### Added
- add golang/dep support for a repeatable build

### Removed
- Removed all usage of go-ceph library.  Simplifies the code and now only
shelling out to 'rbd' executable

### Changed
- Updated docker/go-plugin-helpers and latest sdk api
- Fixed List() to call out to 'rbd ls' instead of relying on in memory per-host list

## [1.5.2] - 2017-08
### Changed
- Remove cron script installation from tpkg

## [1.5.1] - 2016-09-20
### Changed
- Update XFS checks to mount/unmount to clear any disk logs

## [1.5.0] - 2016-09-01
### Added
- Updated plugin to include Capabilities() method and new method
  signatures for Mount(MountRequest) and Unmount(UnmountRequest)
- add xfs filesystem check before mount - if xfs-repair -n returns
  error, Mount operation fails with note to manually repair

## [1.4.1] - 2016-08-05
### Changed
- Bug fix from bad upstream merge: provide openContext and
  shutdownContext

## [1.4.0] - 2016-08-05
### Added
- Merged upstream fixes for pull porcupie/rbd-docker-plugin#7

## [1.3.0] - 2016-08-05
### Added
- Added goroutines and timeouts to all shell commands, hoping to prevent
  propagation of hung external procs to docker daemon

## [1.2.2] - 2016-06-08
### Changed
- Docker is calling Unmount after failed Mount, causing trouble if
  device is still in use by another container (locked by same node).
The workaround / hack is to bail earlier in Unmount API call if rbd
unmap fails with busy device error. This can leave the device usable but
possibly in a funky state (unmounted from host but still mounted and
accessible inside container)
- related to porcupie/rbd-docker-plugin#5

## [1.2.1] - 2016-06-02
### Changed
- When rbd map does not return device name but no error, try default
  device path (/dev/rbd/<pool>/<image>). Issue porcupie/rbd-docker-plugin#4

## [1.2.0] - 2016-06-02
### Added
- Updated to pull in yp-engineering/rbd-docker-plugin v0.9.1.2, which
  includes support for Docker Volume Create options: size, pool, fstype

## [1.1.1] - 2016-04-15
### Changed
- Due to issue with golang 1.6 and strict Host header requirements, we
  cannot use golang 1.6 to compile our plugin since docker never sends
  correct Host header for plugin socket usage.
  * Recompiled with golang 1.5

## [1.1.0] - 2016-04-15
### Added
- added a cron job and check script for the Ceph configs and tpkg update

## [1.0.0] - 2016-04-15
### Changed
- bump major version with deprecated --remove boolean flag
- --remove flag now takes one of three values: ignore, delete or rename
  - ignore will just ignore the docker volume Remove call (new default)
  - delete will destroy the rbd volume on Remove request
  - rename will rename the rbd volume on Remove request, prefixed with `zz_`

## [0.5.0] - 2016-04-13
### Changed
- pulled latest from upstream yp-engineering/rbd-docker-plugin
  - add new docker volume api support (Get, List)
  - use ceph/go-ceph instead of noahdesu/go-ceph
  - use docker/go-plugins-helpers/ instead of calavera/dkvolume

## [0.4.2] - 2016-03-16
### Changed
- Update logrotate config to restart instead of reload
### Added 
- Some new marathon-tester configs for running a test container in
  Marathon/Mesos environment

## [0.4.1] - 2015-12-03
### Changed
- Update systemd service unit to add --config /etc/ceph/ceph.conf 
- Force default config file in main.go to /etc/ceph/ceph.conf

## [0.4.0] - 2015-12-03
### Changed
- Last ditch effort : Update all Plugin RBD functions to use CLI shell
  commands instead of go-ceph library
- Provide command line flag --go-ceph to use go-ceph lib, otherwise default
  now is shell CLI command via `rbd` binary

## [0.3.1] - 2015-11-30
### Changed
- Try to open RBD Image without read-only option (no effect)
- Try to use same client-id for every connection -- not possible in
  go-ceph
- Adding --conf options to external rbd operations (was having micro-osd
  issues)

## [0.3.0] - 2015-11-25
### Changed
- Update go-ceph import to use github.com/ceph/go-ceph instead of
  noahdesu/go-ceph
- Recreate ceph connection and pool context for every operation (don't 
  try to cache them)

## [0.2.2] - 2015-11-19
### Changed
- Disable the reload operation in systemd service unit, having issues
  with go-ceph lib and that operation (panics)
- Update the Image Rename and Remove functions to use go-ceph lib
  instead of shelling out to rbd binary
- Update the tpkg scripts to start the service on installation

## [0.2.1] - 2015-09-11
### Added
- Merged pull request with some RPM scripts for use in generic Redhat EL7 (Thanks Clement Laforet <sheepkiller@cotds.org>)

## [0.2.0] - 2015-08-25
### Changed
- Added micro-osd script for testing Ceph locally

## [0.1.9] - 2015-08-20
### Changed
- Added user ID and options to more shell rbd binary exec commands (Thanks Sébastien Han <seb@redhat.com>)
- Moving version definition from tpkg.yml to version.go
- Better blkid integration (Thanks Sébastien Han <seb@redhat.com>)


utils.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io/ioutil"
	"log"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var (
	defaultShellTimeout = 2 * 60 * time.Second
)

// sh is a simple os.exec Command tool, returns trimmed string output
func sh(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	log.Printf("INFO: sh CMD: %q", cmd)
	out, err := cmd.Output()
	log.Printf("INFO: [out, err]/[%s, %s]", out, err)
	return strings.Trim(string(out), " \n"), err
}

// ShResult used for channel in timeout
type ShResult struct {
	Output string // STDOUT
	Err    error  // go error, not STDERR
}

type ShTimeoutError struct {
	timeout time.Duration
}

func (e ShTimeoutError) Error() string {
	return fmt.Sprintf("Reached TIMEOUT on shell command")
}

// shWithDefaultTimeout will use the defaultShellTimeout so you dont have to pass one
func shWithDefaultTimeout(name string, args ...string) (string, error) {
	return shWithTimeout(defaultShellTimeout, name, args...)
}

// shWithTimeout will run the Cmd and wait for the specified duration
func shWithTimeout(howLong time.Duration, name string, args ...string) (string, error) {
	// duration can't be zero
	if howLong <= 0 {
		return "", fmt.Errorf("Timeout duration needs to be positive")
	}
	// set up the results channel
	resultsChan := make(chan ShResult, 1)
	if isDebugEnabled() {
		log.Printf("DEBUG: shWithTimeout: %v, %s, %v", howLong, name, args)
	}

	// fire up the goroutine for the actual shell command
	go func() {
		out, err := sh(name, args...)
		resultsChan <- ShResult{Output: out, Err: err}
	}()

	select {
	case res := <-resultsChan:
		return res.Output, res.Err
	case <-time.After(howLong):
		return "", ShTimeoutError{timeout: howLong}
	}

	return "", nil
}

// grepLines pulls out lines that match a string (no regex ... yet)
func grepLines(data string, like string) []string {
	var result = []string{}
	if like == "" {
		log.Printf("ERROR: unable to look for empty pattern")
		return result
	}
	like_bytes := []byte(like)

	scanner := bufio.NewScanner(strings.NewReader(data))
	for scanner.Scan() {
		if bytes.Contains(scanner.Bytes(), like_bytes) {
			result = append(result, scanner.Text())
		}
	}
	if err := scanner.Err(); err != nil {
		log.Printf("WARN: error scanning string for %s: %s", like, err)
	}

	return result
}

// regexpLines pulls out lines that match a regexp as group matches
func regexpLines(data string, regexp_s string) [][]string {
	var result = [][]string{}

	r, err := regexp.Compile(regexp_s)
	if err != nil {
		log.Printf("ERROR: unable to compile regexp")
		return result
	}

	scanner := bufio.NewScanner(strings.NewReader(data))
	for scanner.Scan() {
		s := scanner.Text()
		if r.MatchString(s) {
			result = append(result, r.FindStringSubmatch(s))
		}
	}
	if err := scanner.Err(); err != nil {
		log.Printf("WARN: error scanning string for %s: %s", regexp_s, err)
	}

	return result
}

// Linux process management
//ref: https://github.com/kimpettersen/GoProcs/blob/master/src/procs.go
type Process struct {
	Pid        string
	Executable string
}

// List all processes in system
func listProcesses() []Process {
	var processes []Process
	files, err := ioutil.ReadDir("/proc")

	if err != nil {
		log.Fatal("Could not read dir /proc :)")
	}
	var proc Process

	for _, file := range files {
		if _, err := strconv.Atoi(file.Name()); err == nil {

			cmd, err := ioutil.ReadFile("/proc/" + file.Name() + "/cmdline")

			cmdString := strings.Join(strings.Split(string(cmd), "\x00"), " ")

			if err != nil {
				log.Fatal("Can't read file:", err)
			}

			proc = Process{
				Pid:        file.Name(),
				Executable: cmdString,
			}
			processes = append(processes, proc)
		}
	}
	return processes
}

// kill a process
func kill(proc Process, signal string) error {
	if err := exec.Command("kill", "-"+signal, string(proc.Pid)).Start(); err != nil {
		log.Fatal("Couldn't kill ", err)
		return err
	}
	fmt.Printf("Killed %v\n", proc.Executable)
	return nil
}


main.go
// Copyright 2015 YP LLC.
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file.
package main

// Ceph RBD VolumeDriver Docker Plugin, setup config and go

import (
	"errors"
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"os/user"
	"path/filepath"
	"strconv"
	"syscall"

	dkvolume "github.com/docker/go-plugins-helpers/volume"
)

var (
	// Never delete rbd images
	VALID_REMOVE_ACTIONS = []string{"ignore", "rename"}

	// Plugin Option Flags
	versionFlag        = flag.Bool("version", false, "Print version")
	debugFlag          = flag.Bool("debug", false, "Debug output")
	pluginName         = flag.String("name", "rbd", "Docker plugin name for use on --volume-driver option")
	cephUser           = flag.String("user", "admin", "Ceph user")
	cephConfigFile     = flag.String("config", "/etc/ceph/xtao.conf", "Xtao ceph cluster config") // more likely to have config file pointing to cluster
	cephCluster        = flag.String("cluster", "xtao", "xtao ceph cluster")                      // less likely to run multiple clusters on same hardware
	defaultCephPool    = flag.String("pool", "rbd", "Default Ceph Pool for RBD operations")
	pluginDir          = flag.String("plugins", "/run/docker/plugins", "Docker plugin directory for socket")
	rootMountDir       = flag.String("mount", dkvolume.DefaultDockerRootDirectory, "Mount directory for volumes on host")
	logDir             = flag.String("logdir", "/var/log", "Logfile directory")
	canCreateVolumes   = flag.Bool("create", true, "Can auto Create RBD Images")
	defaultImageSizeMB = flag.Int("size", 20*1024, "RBD Image size to Create (in MB) (default: 20480=20GB)")
	defaultImageFSType = flag.String("fs", "xfs", "FS type for the created RBD Image (must be xfs now)")
	useGoCeph          = flag.Bool("go-ceph", false, "Use go-ceph library")
	useNbd             = flag.Bool("use-nbd", true, "Use rbd-nbd to map RBD Image")
)

// setup a validating flag for remove action
type removeAction string

func (a *removeAction) String() string {
	return string(*a)
}

func (a *removeAction) Set(value string) error {
	if !contains(VALID_REMOVE_ACTIONS, value) {
		return errors.New(fmt.Sprintf("Invalid value: %s, valid values are: %q", value, VALID_REMOVE_ACTIONS))
	}
	*a = removeAction(value)
	return nil
}

func contains(vals []string, check string) bool {
	for _, v := range vals {
		if check == v {
			return true
		}
	}
	return false
}

var removeActionFlag removeAction = "ignore"

func init() {
	flag.Var(&removeActionFlag, "remove", "Action to take on Remove: ignore, delete or rename")
	flag.Parse()
}

func socketPath() string {
	return filepath.Join(*pluginDir, *pluginName+".sock")
}

func logfilePath() string {
	return filepath.Join(*logDir, *pluginName+"-docker-plugin.log")
}

func main() {
	if *versionFlag {
		fmt.Printf("%s\n", VERSION)
		return
	}

	logFile, err := setupLogging()
	if err != nil {
		log.Fatalf("FATAL: Unable to setup logging: %s", err)
	}
	defer shutdownLogging(logFile)

	log.Printf("INFO: starting rbd-docker-plugin version %s", VERSION)
	log.Printf("INFO: canCreateVolumes=%q, removeAction=%q", *canCreateVolumes, removeActionFlag)
	log.Printf(
		`INFO: Setting up Ceph Driver for PluginID=%s, cluster=%s, user=%s, pool=%s, mount=%s, 
			config=%s, go-ceph=%s, useNbd=%s`,
		*pluginName,
		*cephCluster,
		*cephUser,
		*defaultCephPool,
		*rootMountDir,
		*cephConfigFile,
		*useGoCeph,
		*useNbd,
	)

	// double check for config file - required especially for non-standard configs
	if *cephConfigFile == "" {
		log.Fatal("FATAL: Unable to use ceph rbd tool without config file")
	}
	if _, err = os.Stat(*cephConfigFile); os.IsNotExist(err) {
		log.Fatalf("FATAL: Unable to find ceph config needed for ceph rbd tool: %s", err)
	}

	// build driver struct -- but don't create connection yet
	d := newCephRBDVolumeDriver(
		*pluginName,
		*cephCluster,
		*cephUser,
		*defaultCephPool,
		*rootMountDir,
		*cephConfigFile,
		*useGoCeph,
		*useNbd,
	)
	if *useGoCeph {
		defer d.shutdown()
	}

	log.Println("INFO: Creating Docker VolumeDriver Handler")
	h := dkvolume.NewHandler(d)

	// setup signal handling after logging setup and creating driver, in order to signal the logfile and ceph connection
	// NOTE: systemd will send SIGTERM followed by SIGKILL after a timeout to stop a service daemon
	signalChannel := make(chan os.Signal, 2) // chan with buffer size 2
	signal.Notify(signalChannel, syscall.SIGTERM, syscall.SIGKILL)
	go func() {
		for sig := range signalChannel {
			//sig := <-signalChannel
			switch sig {
			case syscall.SIGTERM, syscall.SIGKILL:
				log.Printf("INFO: received TERM or KILL signal: %s", sig)
				// close up conn and logs
				if *useGoCeph {
					d.shutdown()
				}
				shutdownLogging(logFile)
				os.Exit(0)
			}
		}
	}()

	u, _ := user.Lookup("root")
	gid, _ := strconv.Atoi(u.Gid)
	err = h.ServeUnix(*pluginName, gid)

	if err != nil {
		log.Printf("ERROR: Unable to create UNIX socket: %v", err)
	}

}

// isDebugEnabled checks for RBD_DOCKER_PLUGIN_DEBUG environment variable
func isDebugEnabled() bool {
	return *debugFlag || os.Getenv("RBD_DOCKER_PLUGIN_DEBUG") == "1"
}

// setupLogging attempts to log to a file, otherwise stderr
func setupLogging() (*os.File, error) {
	// use date, time and filename for log output
	log.SetFlags(log.LstdFlags | log.Lshortfile)

	// setup logfile - path is set from logfileDir and pluginName
	logfileName := logfilePath()
	if !isDebugEnabled() && logfileName != "" {
		logFile, err := os.OpenFile(logfileName, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
		if err != nil {
			// check if we can write to directory - otherwise just log to stderr?
			if os.IsPermission(err) {
				log.Printf("WARN: logging fallback to STDERR: %v", err)
			} else {
				// some other, more extreme system error
				return nil, err
			}
		} else {
			log.Printf("INFO: setting log file: %s", logfileName)
			log.SetOutput(logFile)
			return logFile, nil
		}
	}
	return nil, nil
}

func shutdownLogging(logFile *os.File) {
	// flush and close the file
	if logFile != nil {
		log.Println("INFO: closing log file")
		logFile.Sync()
		logFile.Close()
	}
}

func reloadLogging(logFile *os.File) (*os.File, error) {
	log.Println("INFO: reloading log")
	if logFile != nil {
		shutdownLogging(logFile)
	}
	return setupLogging()
}
